vmm/src/landlock.rs

```diff
@@ -0,0 +1,130 @@
+    #[error("Invalid path")]
+    InvalidPath,
+}
+
+// https://docs.rs/landlock/latest/landlock/enum.ABI.html for more info on ABI
+static ABI: ABI = ABI::V3;
+
+pub(crate) struct LandlockAccess {
+    access: BitFlags<AccessFs>,
+}
+
+impl TryFrom<&str> for LandlockAccess {
+    type Error = LandlockError;
+
+    fn try_from(s: &str) -> Result<LandlockAccess, LandlockError> {
+        if s.is_empty() {
+            return Err(LandlockError::InvalidLandlockAccess(
+                "Access cannot be empty".to_string(),
+            ));
+        }
+
+        let mut access = BitFlags::<AccessFs>::empty();
+        for c in s.chars() {
+            match c {
+                'r' => access |= AccessFs::from_read(ABI),
+                'w' => access |= AccessFs::from_write(ABI),
+                _ => {
+                    return Err(LandlockError::InvalidLandlockAccess(
+                        format!("Invalid access: {c}").to_string(),
+                    ))
+                }
+            };
+        }
+        Ok(LandlockAccess { access })
+    }
+}
+pub struct Landlock {
+    ruleset: RulesetCreated,
+}
+
+impl Landlock {
+    pub fn new() -> Result<Landlock, LandlockError> {
+        let file_access = AccessFs::from_all(ABI);
+
+        let def_ruleset = Ruleset::default()
+            .handle_access(file_access)
+            .map_err(LandlockError::ManageRuleset)?;
+
+        // By default, rulesets are created in `BestEffort` mode. This lets Landlock
+        // to enable all the supported rules and silently ignore the unsupported ones.
+        let ruleset = def_ruleset.create().map_err(LandlockError::ManageRuleset)?;
+
+        Ok(Landlock { ruleset })
+    }
+
+    pub(crate) fn add_rule(
+        &mut self,
+        path: PathBuf,
+        access: BitFlags<AccessFs>,
+    ) -> Result<(), LandlockError> {
+        // path_beneath_rules in landlock crate handles file and directory access rules.
+        // Incoming path/s are passed to path_beneath_rules, so that we don't
+        // have to worry about the type of the path.
+        let paths = vec![path.clone()];
+        let path_beneath_rules = path_beneath_rules(paths, access);
+        self.ruleset
+            .as_mut()
+            .add_rules(path_beneath_rules)
+            .map_err(LandlockError::ManageRuleset)?;
+        Ok(())
+    }
+
+    pub(crate) fn add_rule_with_access(
+        &mut self,
+        path: PathBuf,
+        access: String,
+    ) -> Result<(), LandlockError> {
+        self.add_rule(path, LandlockAccess::try_from(access.as_str())?.access)?;
+        Ok(())
+    }
+
+    pub fn restrict_self(self) -> Result<(), LandlockError> {
+        self.ruleset
+            .restrict_self()
+            .map_err(LandlockError::ManageRuleset)?;
+        Ok(())
+    }
+}
+
+#[test]
+fn test_try_from_perms() {
+    let landlock_access = LandlockAccess::try_from("rw").unwrap();
+    assert!(landlock_access.access == AccessFs::from_read(ABI) | AccessFs::from_write(ABI));
```

Comment:
My comment was to use explicit access rights instead of the `from_*()` helpers to help future reviews when a `from_read(v9)` will return something different than `from_read(v3)`.

Reply:
Gotcha!

I like the idea of highlighting access right changes to reviewers with landlock version bumps. Addressed this in my latest update.